### PR TITLE
Fix NXP-28316 Reindexing without service interruption

### PR DESCRIPTION
**Pull request by bdelbosc**

### CODE REVIEW

**Comment by bdelbosc** [comment]
this is a draft not yet ready for review

**Comment by ataillefer** [comment]
Rebased on 10.10.

**Review by kevinleturc — DISMISSED** [review]
Good work, the comments in the tests helps a lot to understand the case 👍.

File: nuxeo-features/nuxeo-elasticsearch/nuxeo-elasticsearch-core/src/main/java/org/nuxeo/elasticsearch/core/ReindexingMessage.java [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+        IOUtils.write(indexName, out, UTF_8);
+        IOUtils.write(secondWriteIndexName, out, UTF_8);
+        IOUtils.write(String.valueOf(state), out, UTF_8);
+    }
+
+    public static ReindexingMessage deserialize(InputStream in) throws IOException {
+        String id = IOUtils.toString(in, UTF_8);
+        String index = IOUtils.toString(in, UTF_8);
+        String writeIndex = IOUtils.toString(in, UTF_8);
+        ReindexingState state = ReindexingState.valueOf(IOUtils.toString(in, UTF_8));
```
Comment (efge) [inline nuxeo-features/nuxeo-elasticsearch/nuxeo-elasticsearch-core/src/main/java/org/nuxeo/elasticsearch/core/ReindexingMessage.java]: I don't think this does what you want. This is not like an `ObjectInputStream`. The first `IOUtils.toString` will consume all the input stream and return a big string with all that you wrote. What you should do instead is find a delimiter (maybe '\0') and write a single string with the delimiter, and on read do a split. See `CacheServiceImpl.CacheInvalidation` for an example.

**Review by efge — CHANGES_REQUESTED** [review]

File: nuxeo-features/nuxeo-elasticsearch/nuxeo-elasticsearch-core/src/main/java/org/nuxeo/elasticsearch/core/ReindexingMessage.java (hunk 1 above)
Comment (bdelbosc) [reply nuxeo-features/nuxeo-elasticsearch/nuxeo-elasticsearch-core/src/main/java/org/nuxeo/elasticsearch/core/ReindexingMessage.java]: good catch, thanks will add a unit test

File: nuxeo-features/nuxeo-elasticsearch/nuxeo-elasticsearch-core/src/main/java/org/nuxeo/elasticsearch/core/ReindexingState.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+ */
+package org.nuxeo.elasticsearch.core;
+
+/**
+ * @since 2021.11
+ */
+public enum ReindexingState {
+    START("START"), END("END"), ABORT("ABORT");
+
+    private String name;
```
Comment (efge) [inline nuxeo-features/nuxeo-elasticsearch/nuxeo-elasticsearch-core/src/main/java/org/nuxeo/elasticsearch/core/ReindexingState.java]: If you store exactly the enum name I don't think you need a private field.

**Review by efge — DISMISSED** [review]

**Review by efge — DISMISSED** [review]

**Review by kevinleturc — DISMISSED** [review]

File: nuxeo-runtime/nuxeo-runtime-pubsub/src/main/java/org/nuxeo/runtime/pubsub/StreamPubSubProvider.java [hunk 3]
```diff
@@ -142,7 +142,12 @@ public void run() {
         }
     }
 
-    protected String getNodeId() {
+    /**
+     * Returns the cluster node id.
+     *
+     * @since 2021.8
```
Comment (ataillefer) [inline nuxeo-runtime/nuxeo-runtime-pubsub/src/main/java/org/nuxeo/runtime/pubsub/StreamPubSubProvider.java:148]: 2021.8?

**Review by ataillefer — DISMISSED** [review]

**Comment by ataillefer** [comment]
Rebased on 10.10.

**Review by ataillefer — APPROVED** [review]

File: nuxeo-runtime/nuxeo-runtime-pubsub/src/main/java/org/nuxeo/runtime/pubsub/StreamPubSubProvider.java (hunk 3 above)
Comment (bdelbosc) [reply nuxeo-runtime/nuxeo-runtime-pubsub/src/main/java/org/nuxeo/runtime/pubsub/StreamPubSubProvider.java:148]: will fix the since on lts PR

**Review by kevinleturc — APPROVED** [review]

**Pull request merged**